src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1319,6 +1321,150 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+    dout(5) << ": cannot proceeed with sync: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  BOOST_SCOPE_EXIT_ALL( (this)(&fh) ) {
+    post_sync_close_handles(fh);
+  };
+
+  // record that we are going to "dirty" the data under this directory root
+  auto snap_id_str{stringify(current.second)};
+  r = ceph_setxattr(m_remote_mount, dir_root.c_str(), "ceph.mirror.dirty_snap_id",
+                    snap_id_str.c_str(), snap_id_str.size(), 0);
+  if (r < 0) {
+    derr << ": error setting \"ceph.mirror.dirty_snap_id\" on dir_root=" << dir_root
+         << ": " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  ceph_snapdiff_info sd_info;
+  ceph_snapdiff_entry_t sd_entry;
+  std::string epath = ".";
+  std::string npath = "", nabs_path = "", nname = "";
+  int ntype = 0;
+
+  //the queue of SyncDiffEntry items (directories) to be synchronized
+  std::queue<SyncDiffEntry> sync_queue;
+
+  //start with initial/default entry
+  strncpy(sd_entry.dir_entry.d_name, ".", sizeof(sd_entry.dir_entry.d_name));
+  sd_entry.dir_entry.d_type = 4;
+  sync_queue.emplace(SyncDiffEntry(sd_entry, epath));
+
+  while (!sync_queue.empty()) {
+    if (should_backoff(dir_root, &r)) {
+      dout(0) << ": backing off r=" << r << dendl;
+      break;
+    }
+
+    dout(20) << ": " << sync_queue.size() << " entries in stack" << dendl;
+    const auto &queue_entry = sync_queue.front();
+    epath = queue_entry.path;
+    dout(20) << ": syncing entry, path=" << epath << dendl;
+    r = ceph_open_snapdiff(m_local_mount, dir_root.c_str(), epath.c_str(),
+                           stringify((*prev).first).c_str(), current.first.c_str(), &sd_info);
+    if (r != 0) {
+      derr << ": failed to open snapdiff, r=" << r << dendl;
+      return r;
+    }
+    while (0 < (r = ceph_readdir_snapdiff(&sd_info, &sd_entry))) {
+      if (r < 0) {
+        derr << ": failed to read directory=" << epath << dendl;
+        continue;
+      }
+
+      //New entry found
+      nname = sd_entry.dir_entry.d_name;
+      ntype = sd_entry.dir_entry.d_type;
+      // create path for the newly found entry
+      npath = entry_diff_path(epath, nname);
+      nabs_path = entry_diff_path(dir_root, npath);
+
+      struct ceph_statx cstx;
+      r = ceph_statx(sd_info.cmount, nabs_path.c_str(), &cstx,
+                     CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                     CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                     AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+
+      if (ntype == 4) { // is a directory
+        if (nname == "." || nname == "..")
+          continue;
+        if (r < 0) {
+          // can't stat, so it's a deleted entry. Then remove from the remotefs.
+          r = ceph_rmdir(m_remote_mount, nabs_path.c_str());
+          if (r < 0 && r != -ENOTEMPTY) {
+            derr << ": failed to remove directory=" << nabs_path << dendl;
+            break;//avoid half-baked
+          } else if (r == -ENOTEMPTY) {
```

Comment:
Why not call `cleanup_remote_dir()` directly before calling `ceph_rmdir()`?

Reply:
Will check that out, if it works.